Aluno: Adriel Ferreira Trajano

Matrícula: 20220071160

In [ ]:
pip install nbconvert


In [2]:
import numpy as np

Questão 1 - letra d

$ B1 = \{
v1 = \begin{bmatrix} 0 & 1 \\ 1 & 0 \end{bmatrix}, 
v2 = \begin{bmatrix} 1 & 1 \\ 0 & 1 \end{bmatrix}, 
v3 = \begin{bmatrix} 1 & 0 \\ 1 & 1 \end{bmatrix}, 
v4 = \begin{bmatrix} 1 & 1 \\ 1 & 0 \end{bmatrix}
\}
$



$[v]_{B1} = (1, 1, 1, 1)$

In [3]:
v1 = np.array([[0,1],[1,0]], dtype=float)
v2 = np.array([[1,1],[0,1]], dtype = float)
v3 = np.array([[1,0],[1,1]], dtype = float)
v4 = np.array([[1,1],[1,0]], dtype = float)
B1 = ([v1,v2,v3,v4])

In [4]:
i=1
for v in B1:
    print(f"v{i} = {v}\n")
    i+=1

v1 = [[0. 1.]
 [1. 0.]]

v2 = [[1. 1.]
 [0. 1.]]

v3 = [[1. 0.]
 [1. 1.]]

v4 = [[1. 1.]
 [1. 0.]]



* Ortogonalização

\begin{align*}
w_1 & = v_1 \\
w_2 & = \frac{\langle v_2, w_1 \rangle}{\|w_1\|^2} \vec{w_1} \\
w_3 & = v_3 - \frac{\langle v_3, w_1 \rangle}{\|w_1\|^2} \vec{w_1} - \frac{\langle v_3, w_2 \rangle}{\|w_2\|^2} \vec{w_2} \\
w_4 & = v_4 - \frac{\langle v_4, w_1 \rangle}{\|w_1\|^2} \vec{w_1} - \frac{\langle v_4, w_2 \rangle}{\|w_2\|^2} \vec{w_2} - \frac{\langle v_4, w_3 \rangle}{\|w_3\|^2} \vec{w_3}
\end{align*}


In [5]:
ortho_basis = [] #base ortogonal

Obs: o código abaixo faz o cálculo dos vetores wk, que formam a base ortogonal

In [6]:
# w1 é igual a v1
w1 = v1
ortho_basis.append(w1)

# Calculando w2
proj_v2_w1 = (np.trace(np.dot(w1.T, v2)) / np.trace(np.dot(w1.T, w1))**2) * w1
w2 = v2 - proj_v2_w1
ortho_basis.append(w2)

# Calculando w3
proj_v3_w2 = (np.trace(np.dot(w2.T, v3)) / np.trace(np.dot(w2.T, w2))**2) * w2
proj_v3_w1 = (np.trace(np.dot(w1.T, v3)) / np.trace(np.dot(w1.T, w1))**2) * w1
w3 = v3 - proj_v3_w2 - proj_v3_w1
ortho_basis.append(w3)

# Calculando w4
proj_v4_w3 = (np.trace(np.dot(w3.T, v4)) / np.trace(np.dot(w3.T, w3))**2) * w3
proj_v4_w2 = (np.trace(np.dot(w2.T, v4)) / np.trace(np.dot(w2.T, w2))**2) * w2
proj_v4_w1 = (np.trace(np.dot(w1.T, v4)) / np.trace(np.dot(w1.T, w1))**2) * w1
w4 = v4 - proj_v4_w3 - proj_v4_w2 - proj_v4_w1
ortho_basis.append(w4)

In [7]:
for i, w in enumerate(ortho_basis, 1):
    print(f"w{i} = {w}")

w1 = [[0. 1.]
 [1. 0.]]
w2 = [[ 1.    0.75]
 [-0.25  1.  ]]
w3 = [[ 0.74603175 -0.44047619]
 [ 0.81349206  0.74603175]]
w4 = [[ 0.5669595   0.46388486]
 [ 0.31959489 -0.4330405 ]]


* Vetores da base ortogonalizada:

$ w_1 = \begin{bmatrix} 0 & 1 \\ 1 & 0 \end{bmatrix} $

$ w_2 = \begin{bmatrix} 1 & 0.75 \\ -0.25 & 1 \end{bmatrix} $

$ w_3 = \begin{bmatrix} 0.74603175 & -0.44047619 \\ 0.81349206 & 0.74603175 \end{bmatrix} $

$ w_4 = \begin{bmatrix} 0.5669595 & 0.46388486 \\ 0.31959489 & -0.4330405 \end{bmatrix} $


* Vamos agora achar a base ortonormal B2

In [8]:
base_ortonormal = []

Aqui calculamos os vetores w_til_k, que formam a base ortonomal

In [9]:
for i in range(len(ortho_basis)):
    w = ortho_basis[i]
    w_til = w/np.trace(np.dot(w.T,w))
    base_ortonormal.append(w_til) 

In [10]:
w_til_1, w_til_2, w_til_3, w_til_4 = base_ortonormal

In [13]:
for i, w in enumerate(base_ortonormal, 1):
    print(f"w_til_{i} = {w}")

w_til_1 = [[0.  0.5]
 [0.5 0. ]]
w_til_2 = [[ 0.38095238  0.28571429]
 [-0.0952381   0.38095238]]
w_til_3 = [[ 0.37890494 -0.22371515]
 [ 0.41316762  0.37890494]]
w_til_4 = [[ 0.68614476  0.56140194]
 [ 0.38677959 -0.52407354]]


* Base ortonormalizada

$ 
B2 = \{
\widetilde{w}_1 = \begin{bmatrix} 0 & 0.5 \\ 0.5 & 0 \end{bmatrix},$

$\widetilde{w}_2 = \begin{bmatrix} 0.38095238 & 0.28571429 \\ -0.0952381 & 0.38095238 \end{bmatrix},$

$ \widetilde{w}_3 = \begin{bmatrix} 0.37890494 & -0.22371515 \\ 0.41316762 & 0.37890494 \end{bmatrix}, $

$ \widetilde{w}_4 = \begin{bmatrix} 0.68614476 & 0.56140194 \\ 0.38677959 & -0.52407354 \end{bmatrix} 
 \}$


* Vamos agora fazer a matriz mudança da base B1 para a base B2 e verificar se a mesma é triangular.

In [17]:
B2 = base_ortonormal
B2

[array([[0. , 0.5],
        [0.5, 0. ]]),
 array([[ 0.38095238,  0.28571429],
        [-0.0952381 ,  0.38095238]]),
 array([[ 0.37890494, -0.22371515],
        [ 0.41316762,  0.37890494]]),
 array([[ 0.68614476,  0.56140194],
        [ 0.38677959, -0.52407354]])]

Concatenação das matrizes da base B2

In [19]:
P = np.concatenate((w_til_1, w_til_2, w_til_3, w_til_4), axis=1)

print(P)

[[ 0.          0.5         0.38095238  0.28571429  0.37890494 -0.22371515
   0.68614476  0.56140194]
 [ 0.5         0.         -0.0952381   0.38095238  0.41316762  0.37890494
   0.38677959 -0.52407354]]


$$ P = \begin{bmatrix}
0 & 0.5 & 0.38095238 & 0.37890494 & 0.68614476 \\
0.5 & 0 & 0.28571429 & -0.22371515 & 0.56140194 \\
0.5 & 0 & -0.0952381 & 0.41316762 & 0.38677959 \\
0 & 0.5 & 0.38095238 & 0.37890494 & -0.52407354 \\
\end{bmatrix} $$


Não é uma matriz triangular.